File: Easy/2287_Rearrange_Characters_to_Make_Target_String/Solution.java



public class Solution {
    public int rearrangeCharacters(String s, String target) {
        int[] freqS = new int[26];
        int[] freqT = new int[26];

        for (char c : s.toCharArray()) {
            freqS[c - 'a']++;
        }

        for (char c : target.toCharArray()) {
            freqT[c - 'a']++;
        }

        int minNum = Integer.MAX_VALUE;

        for (int i = 0; i < 26; i++) {
            if (freqT[i] > 0) {
                minNum = Math.min(minNum, freqS[i] / freqT[i]);
            }
        }
        return minNum;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        String s1 = "ilovecodingonleetcode", target1 = "code";
        System.out.println(solution.rearrangeCharacters(s1, target1));

        // test cases 2cout << solution.rearrangeCharacters(s2, target2) << endl;
        String s2 = "abcba", target2 = "abc";
        System.out.println(solution.rearrangeCharacters(s2, target2));

        // test cases 3
        String s3 = "abbaccaddaeea", target3 = "aaaaa";
        System.out.println(solution.rearrangeCharacters(s3, target3));
        
        
    }
}

File: Easy/2287_Rearrange_Characters_to_Make_Target_String/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    int rearrangeCharacters(string s, string target) {
        unordered_map<char, int> freqS;
        unordered_map<char, int> freqT;

        for (char c : s) {
            freqS[c]++;
        }

        for (char c : target) {
            freqT[c]++;
        }

        int minNum = INT_MAX;

        for (const auto& pair : freqT) {
            minNum = min(minNum, freqS[pair.first] / pair.second);
        }

        return minNum;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    string s1 = "ilovecodingonleetcode", target1 = "code";
    cout << solution.rearrangeCharacters(s1, target1) << endl;

    // test cases 2
    string s2 = "abcba", target2 = "abc";
    cout << solution.rearrangeCharacters(s2, target2) << endl;

    // test cases 3
    string s3 = "abbaccaddaeea", target3 = "aaaaa";
    cout << solution.rearrangeCharacters(s3, target3) << endl;

    return 0;
}